main.nf

```diff
@@ -243,68 +251,61 @@ workflow analysis_reference_based {
         deseq2_script_refactor_reportingtools_table
         deseq2_script_improve_deseq_table
 
-  main:
-    //trim
-    fastp(illumina_input_ch)
-
-    //remove rRNA
-    sortmerna(fastp.out.sample_trimmed, sortmerna_db)
-
-    //map
-    hisat2(sortmerna.out.no_rna_fastq, hisat2_index)
-
-    //count
-    featurecounts(hisat2.out.sample_bam, annotation)
-
-    //prepare annotation for R input
-    prepare_annotation_gene_rows(annotation)
-    prepare_annotation(annotation)
-
-    //defs
-    deseq2_comparisons = dge_comparisons_input_ch
-        .map { it.join(":") }
-        .map { "\"${it}\"" }
-        .collect()
-        .map { it.join(",") }
-
-    featurecounts.out.counts
-        .join( annotated_reads
-                .map{row -> [row[0], row[-2]]} 
-        )
-        .fork{ tupel ->
-          name: tupel[0]
-          count_file: tupel[1]
-          condition: tupel[2]
-        }
-        .set { tpm }
-
-    tpm_filter(tpm.name.toSortedList(), tpm.count_file.toSortedList(), tpm.condition.toSortedList())
-
-    tpm_filter.out.samples
-        .flatMap()
-        .join( annotated_reads
+    main:
+        // trim with fastp
+        fastp(illumina_input_ch)
+
+        // remove rRNA with SortmeRNA
+        sortmerna(fastp.out.sample_trimmed, sortmerna_db)
+
+        // map with HISAT2
+        hisat2(sortmerna.out.no_rna_fastq, hisat2_index)
+
+        // count with featurecounts
+        featurecounts(hisat2.out.sample_bam, annotation)
+
+        // prepare annotation for R input
+        prepare_annotation_gene_rows(annotation)
+        prepare_annotation(annotation)
+
+        // filter by TPM value
+        featurecounts.out.counts
+            .join( annotated_reads
+                    .map{row -> [row[0], row[-2]]} 
+            )
+            .multiMap{ tupel ->
```

Comment:
At some point you have to explain to me this channel magic happening here :)

Reply:
The aim here is to get the condition of a sample (`tuple(sample_name,count_file)`) emitted from the featurecounts process.

`annotated_reads` is basically the input.csv: `tuple(sample_name, read_file, condition, patient)` for each sample.

From `annotated_reads` I just need `sample_name` and `condition`:
```
annotated_reads
        .map{row -> [row[0], row[-2]]} 
```
This I combine with the emitted `tuple(sample_name,count_file)` from featurecounts:
```
featurecounts.out.counts
        .join( annotated_reads
                .map{row -> [row[0], row[-2]]} 
        )
```
[join](https://www.nextflow.io/docs/latest/operator.html#join) combines two channels based on a matching key - by default the first element emitted - in our case `sample_name`.

Now we have for each sample `tuple(sample_name, count_file, condition)`.

`multiMap` splits the three values into separate channels, so that I can `collect` them later on.

Reply:
Thanks, especially the multiMap was not clear to me.